### ISSUE TITLE
Incorrect/Random warning about GraphQL query in non-page component

### ISSUE THREAD

**Issue by dotlouis** [gatsbyjs/gatsby#12214]
## Description

Hello,
We're encountering an issue in which our Gatsby builds (`develop` and `build` alike) are throwing warnings irregularly about GraphQL queries in non-page components.
This should not happen because the query in non-page component is actually a page component in a `template/` directory which are then instanciated in _gatsby-node.js_.
Besides, if the warning is thrown, the query is not run and the page fails at runtime, not having access to its data.

Here are one of the warnings:
```
warning: The GraphQL query in the non-page component "/home/mario/dev/mmh/website/website-front/src/templates/category-zip.jsx" will not be run.
```
What's incomprehensible is that those warning sometimes are triggered, sometimes not. Without anything changing between two builds.
This is a really weird one and we've been skimming the documentation, as well as multiple github issues and stack overflow questions without finding a satisfactory track to get onto.

As this is happening inconstantly, the build step becomes frustrating as we never know if it's going to go well. Has anyone had the same issue before? In which direction should we look for?

### Steps to reproduce

Sadly we haven't been able to reproduce the issue as it is happening randomly on our end. (It happens on local and netlify). Sometimes clearing the cache (`rm -rf .cache/`) works, often not.

We can provide termporary access to the build and logs if needed

### Expected result

The warning should not go off and the pages should be populated with the queries' result.

### Actual result

Warning like this one are sometimes triggered for every pages in the `template/` directory.
```
warning: The GraphQL query in the non-page component "/home/mario/dev/mmh/website/website-front/src/templates/category-zip.jsx" will not be run.
```

### Environment
```
 System:
    OS: macOS 10.14.3
    CPU: (4) x64 Intel(R) Core(TM) i7-7660U CPU @ 2.50GHz
    Shell: 3.0.2 - /usr/local/bin/fish
  Binaries:
    Node: 11.10.0 - /usr/local/bin/node
    Yarn: 1.13.0 - /usr/local/bin/yarn
    npm: 6.8.0 - /usr/local/bin/npm
  Languages:
    Python: 2.7.15 - /usr/local/bin/python
  Browsers:
    Chrome: 72.0.3626.119
    Firefox: 64.0.2
    Safari: 12.0.3
  npmPackages:
    gatsby: 2.1.2 => 2.1.2 
    gatsby-cli: 2.4.10 => 2.4.10 
    gatsby-image: 2.0.29 => 2.0.29 
    gatsby-plugin-copy: 0.0.1 => 0.0.1 
    gatsby-plugin-emotion: 2.0.7 => 2.0.7 
    gatsby-plugin-mailchimp: https://github.com/benjaminhoffman/gatsby-plugin-mailchimp.git#gatsby-v2 => 3.3.0 
    gatsby-plugin-manifest: 2.0.17 => 2.0.17 
    gatsby-plugin-netlify: 2.0.8 => 2.0.8 
    gatsby-plugin-react-helmet: 3.0.6 => 3.0.6 
    gatsby-plugin-robots-txt: 1.4.0 => 1.4.0 
    gatsby-plugin-sharp: 2.0.20 => 2.0.20 
    gatsby-plugin-sitemap: 2.0.5 => 2.0.5 
    gatsby-plugin-subfont: 1.0.3 => 1.0.3 
    gatsby-plugin-webpack-bundle-analyzer: 1.0.4 => 1.0.4 
    gatsby-source-contentful: 2.0.29 => 2.0.29 
    gatsby-transformer-sharp: 2.1.13 => 2.1.13 
  npmGlobalPackages:
    gatsby-contentful-emojis: 0.2.3
```

**Comment by DSchau** [comment]
This will be a challenging one to debug if it only happens intermittently.

What would help most debug this would be:

1. A reproduction/repo that we can take a look at
1. Clear steps that illustrate the reproduction

In the absence of 2) I can attempt to reproduce with 1) but it'll be a bit more challenging. If you can provide even reasonably consistent steps you've taken to illustrate the problem, that would help.

Thanks!

**Comment by dotlouis** [comment]
An update after a lucky day,
We've updated our gatsby version from 2.1.2 to 2.1.23 and we have been seeing a new warning in our build log:
`warning Action createPage was called outside of its expected asynchronous lifecycle createPages in default-site-plugin.
Ensure that you return a Promise from createPages and are awaiting any asynchronous method invocations (like graphql or http requests).
For more info and debugging tips: see https://gatsby.dev/sync-actions`

After following the given link and inspecting our code, we have constated that, at one point, we changed the return value of our `exports.createPages` lifecycle function. It was no longer returning a promise. Thus a race-condition was created and that explains the intermittent success/failure we had.

the commit that introduced the life-saver warning for info: https://github.com/gatsbyjs/gatsby/commit/258b65518b3c6e12a81b9167e29b475287b36345

I'll close this issue. Thanks to @pieh for his work, and you guys for the support.
